File: ch-1/exer-1/sol_1/main/main.cpp
#include <vector>
#include <string>
#include <memory>
#include <thread>
#include <atomic>
#include <chrono>
#include <mutex>
#include "Chopstick.h"
#include "Person.h"
#include "Table.h"

std::mutex g;

void what_do(std::atomic<Table*>& table, std::string name, int offset) {
    for (auto i = 0; i < 3; i++) {
        auto table_ptr = table.load();
        g.lock();
        table_ptr->eat(name);
        g.unlock();
        table.store(table_ptr);

        std::this_thread::sleep_for(std::chrono::milliseconds(offset));
        if (table.load()->get(name)->is_eating) {
            table_ptr = table.load();
            g.lock();
            table_ptr->think(name);
            g.unlock();
            table.store(table_ptr);
            return;
        }
    }
}

int main() {
    std::vector<std::string> v {"Burke",
            "Darwin",
            "Descartes",
            "Plato"};
    std::atomic<Table*> tbl = new Table(v);

    std::thread t1(what_do, std::ref(tbl), "Burke", 1);
    std::thread t2(what_do, std::ref(tbl), "Darwin", 2);
    std::thread t3(what_do, std::ref(tbl), "Descartes", 1.5);
    std::thread t4(what_do, std::ref(tbl), "Plato", 1.25);

    t1.join();
    t2.join();
    t3.join();
    t4.join();
    
    tbl.load()->print_status();
    delete tbl;
    
    return 0;
}

File: ch-1/exer-1/sol_1/java/Main.java
import include.*;
import java.util.ArrayList;

public class Main {
    private static Object lock1 = new Object();
    private static Object lock2 = new Object();
    private static class ThreadPerson implements Runnable {
        private String name;
        private int time_offset;
        private Table table;
        
        public ThreadPerson(Table table, String name, int time_offset) {
            this.table = table;
            this.name = name;
            this.time_offset = time_offset;
        }
        public void run() {
            for (int iter = 0; iter < 3; iter++) {
                synchronized(lock1) {
                    table.eat(name);
                }
                try {
                    Thread.sleep(time_offset);
                } catch (Exception e) {
                    System.err.println("Fatal error: Interrupted Thread");
                    return;
                }
                if (table.get(name).is_eating) {
                    synchronized(lock2) {
                        table.think(name);
                    }
                    return;
                }
            }
        }
    }

    public static void main(String[] args) 
        throws InterruptedException {
        String[] names = {"Burke", "Darwin", "Descartes", "Plato"};
        Table table = new Table(names);
        ArrayList<Thread> lst = new ArrayList<>();

        for (int i = 0; i < 4; i++) {
            Thread thrd = new Thread(new ThreadPerson(table, names[i], 1 + i));
            lst.add(thrd);
        }
        for (Thread thrd : lst) {
            thrd.start();
        }
        for (Thread thrd : lst) {
            thrd.join();
        }
        System.out.println("Done!");
        table.print();
    }
}
